src/lib/utils/debounce.ts

```diff
@@ -0,0 +1,39 @@
+// Debounce the function call to the trailing edge.
```

Comment:
I would still need to test this in detail, but I think this is the typed variant of the current code:


```ts
type Func = (...args: any[]) => any;

type DebouncedFunction<TFunction extends Func> = (
  this: ThisParameterType<TFunction>,
  ...args: Parameters<TFunction>
) => Promise<ReturnType<TFunction>>;

// Debounce the function call to the trailing edge.
export function debounce<TFunction extends Func>(
  func: TFunction,
  wait: number
): DebouncedFunction<TFunction> {
  let lastTimeout: ReturnType<typeof setTimeout> | null = null;
  return function(...args) {
    const that = this;
    return new Promise(resolve => {
      clearTimeout(lastTimeout!);
      lastTimeout = setTimeout(() => {
        lastTimeout = null;
        resolve(func.apply(that, args));
      }, wait);
    });
  };
}

// Debounced function will return a promise which will be either resolved or rejected.
export function debounceAsync<TFunction extends Func>(
  func: TFunction,
  wait: number
): DebouncedFunction<TFunction> {
  let lastTimeout: ReturnType<typeof setTimeout> | null = null;
  let lastReject: (reason?: any) => void | undefined;
  return function(...args) {
    const that = this;
    return new Promise((resolve, reject) => {
      if (typeof lastReject === 'function') {
        lastReject();
      }
      lastReject = reject;
      clearTimeout(lastTimeout!);
      lastTimeout = setTimeout(() => {
        lastTimeout = null;
        Promise.resolve(func.apply(that, args))
          .then(resolve)
          .catch(reject);
      }, wait);
    });
  };
}
```

Reply:
https://github.com/algolia/instantsearch.js/pull/4581/commits/3df3e8102bf72e310cd1bfbe4aa3823c30d24dbe

Thanks! It seems to work fine.
I made a little change to the implementation of `debounce` because it doesn't have to return a Promise. It was mistakenly left there where I started to write debounceAsync.